botorch/acquisition/proximal.py

```diff
@@ -0,0 +1,128 @@
+#!/usr/bin/env python3
+# Copyright (c) Facebook, Inc. and its affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+r"""
+A wrapper around AquisitionFunctions to fix certain features for optimization.
+This is useful e.g. for performing contextual optimization.
+"""
+
+from __future__ import annotations
+
+import torch
+from botorch.acquisition.analytic import AnalyticAcquisitionFunction
+from botorch.utils import t_batch_mode_transform
+from torch import Tensor
+from torch.nn import Module
+
+
+class ProximalAcquisitionFunction(AnalyticAcquisitionFunction):
+    """A wrapper around AquisitionFunctions to add proximal weighting of the acquisition function. Acquisition
+    function is weighted via a squared exponential centered at the last training point, with varying lengthscales
+    corresponding to 'proximal_weights'. Can only be used with single batch analytical acquisition functions.
+
+    Example:
+        >>> model = SingleTaskGP(train_X, train_Y)  # d = 2
+        >>> EI = ExpectedImprovement(model, best_f=0.0)
+        >>> proximal_weights = torch.ones(d)
+        >>> EI_proximal = ProximalAcquisitionFunction(EI, proximal_weights)
+        >>> eip = EI_proximal(test_X)  # d' = 3
+    """
+
+    def __init__(
+        self,
+        acq_function: AnalyticAcquisitionFunction,
+        proximal_weights: Tensor,
+    ) -> None:
+        r"""Derived Acquisition Function by weighted by proximity to recently observed point.
+
+        Args:
+            acq_function: The base acquisition function, operating on input
+                tensors of feature dimension `d`.
+            proximal_weights: Tensor used to bias locality along each axis, should be in the shape of ('d',).
+
+        """
+        Module.__init__(self)
+        self.acq_func = acq_function
+
+        self.register_buffer("proximal_weights", proximal_weights)
+
+        # check to make sure that weights match the training data shape
+        assert (
+            self.proximal_weights.shape[0]
+            == self.acq_func.model.train_inputs[0][-1].shape[-1]
+        )
+
+    @t_batch_mode_transform(expected_q=1, assert_output_shape=False)
+    def forward(self, X: Tensor):
+        r"""Evaluate base acquisition function with proximal weighting.
+
+        Args:
+            X: Input tensor of feature dimension `d` .
+
+        Returns:
+            Base acquisition function evaluated on tensor `X` multiplied by a squared exponenital centered at the last
+            training point and is weighted according to the weighting of each dimension.
+        """
+        d = self.proximal_weights.shape[0]
+        last_X = self.acq_func.model.train_inputs[0][-1].reshape(1, 1, -1)
```

Comment:
Strictly, this won't work for all models as the basic `Model` API does not enforce a model to have `train_inputs`. Can you add a check to the constructor that raises an `UnsupportedError` with an informative error message if `acqf.model` does not have a `train_inputs` attribute? You should probably also raise an error if the "batch shape", i.e. `train_inputs[0][-1].shape[:-3]` is not empty; as this will not work with batched models. If you want to make it work with batched models this `reshape` should be modified here.